README.rst
Svnwrap - a wrapper script for Subversion
=========================================

.. image:: https://travis-ci.org/drmikehenry/svnwrap.svg
    :target: https://travis-ci.org/drmikehenry/svnwrap

.. image:: https://img.shields.io/pypi/v/svnwrap.svg
    :target: https://pypi.python.org/pypi/svnwrap

.. image:: https://img.shields.io/pypi/status/svnwrap.svg
    :target: https://pypi.python.org/pypi/svnwrap

.. image:: https://img.shields.io/pypi/pyversions/svnwrap.svg
    :target: https://pypi.python.org/pypi/svnwrap/

Svnwrap extends the functionality of svn, the command-line interface for the
Subversion version control system.  Typically, the user will define a shell
alias for ``svn`` that invokes ``svnwrap``.  In this way, operations like ``svn
status`` will be handled by svnwrap instead of svn.  For the Bash shell, the
following alias could be placed in ``~/.bashrc`` to make the ``svn`` command
invoke ``svnwrap``::

  alias svn='svnwrap'

Features
--------

- Suppression of noisy output from certain operations such as ``svn status``
  (especially beneficial when using svn:externals).

- Color highlighting of status, diff, and other outputs.

- Extended "diff" operations (including integration with kdiff3).

- Configurable URL aliases of the form ``//alias`` that map to arbitrary URL
  prefixes.  Configuring the alias ``proj`` to be ``http://server/Project``
  would make the following commands identical::

    svn checkout //proj/some/path
    svn checkout http://server/Project/some/path

- URL mapping using keywords that takes advantage of context within a working
  copy.  So, for example, in a working copy checked out from
  http://server/Project/trunk/some/path, creating a tag could be done via::

    svn copy tr: tag:tagname

  The working copy's URL (http://server/Project/trunk/some/path) is used as
  context to allow the ``tr:`` keyword to extract everything before the
  "middle" part (``/trunk`` in this case) and append ``/trunk``.  The
  ``tag:`` keyword behaves similarly, but appends ``/tags`` instead of
  ``/trunk``.  Thus, the above ``svn copy`` operation is equivalent to::

    svn copy http://server/Project/trunk http://server/Project/tags/tagname

  Switching or merging a tag is shortened as well::

    svn switch tag:tagname/some/path
    svn merge tag:tagname/some/path

- URL adjustment for certain commands.  URL suffixes like ``/some/path`` may
  often be omitted during a ``switch`` or ``merge`` operation because svnwrap
  can infer the suffix from the context of the current checkout.  For example,
  when executed in a working copy checked out from
  http://server/Project/trunk/some/path, the following are pairs of equivalent
  commands::

    svn switch tag:tagname/some/path
    svn switch tag:tagname

    svn merge tag:tagname/some/path
    svn merge tag:tagname

- Additional new subcommands such as:

  - ``svn branch`` for creating branches.

  - ``svn ee`` for editing svn:externals.

- See built-in help for more details::

    svnwrap helpwrap


Configuration
-------------

Svnwrap is configured via a configuration file, typically at one of these
locations::

  # On Unix:
  ~/.config/svnwrap/config

  # On Windows:
  %APPDATA%\svnwrap\config

On first invocation of svnwrap, the config file will be created with a commented
skeleton.

Caveats
-------

- On occasion, the ``svn`` client needs to invoke an editor (e.g., to provide a
  commit message).  Without svnwrap, this works fine because the stdout and
  stderr of ``svn`` are connected to a terminal.  But to created prettied output
  and to capture error messages, svnwrap uses pipes to capture stdout and stderr
  from ``svn``, which makes some console-based editors unable to function
  correctly.  To work around this problem, svnwrap tries to determine which
  editor ``svn`` would invoke, then it sets the ``SVN_EDITOR`` environment
  variable to run some Python code before invoking the original editor.  When
  ``svn`` executes this code, stdout and stderr are first reconnected to the
  console, then the original editor is invoked.  Svnwrap looks in most of the
  places where an editor might be configured, but it checks only per-user
  environment variables and config files.  It will not check any registry
  settings on Windows, nor will it check any system-wide configuration files.
  To overcome this limitation, set the ``SVN_EDITOR`` environment variable to
  your preferred editor settings.

License
-------

Svnwrap is available under the terms of the MIT license; see LICENSE.rst file
for more details.

Changes
-------

See the file CHANGES.rst for details on changes to svnwrap.


CHANGES.rst
*******
Changes
*******

Significant changes to svnwrap (newest changes first).

Version 0.8.2
=============

- Support universal wheels.

- Small changes: update copyright dates, tweak comment wording, ignore
  ``__pycache__/`` directories.

Version 0.8.1
=============

- Support new core metadata file format (version 2.1), which allows the
  "long_description" (i.e., README.rst text) to be stored as the metadata "body"
  instead of as the "Description" header).  This fixes ``svnwrap readme`` in
  cases where the new metadata format is being used.

Version 0.8.0
=============

- Update Python-related tooling (black, mypy with type-hinting comments, pytest,
  tox, coverage).

- Reorganize source tree to use ``src/`` subdirectory.

- Restructure code to reduce cyclomatic complexity.

- Support Python 3.8.

- Add "readme" command to display the top-level README.rst.

- Update the list of known switches (now supporting through Subversion 1.13).

Version 0.7.14
==============

- Ensure colors comes back as a list (thanks to John Szakmeister).  This fixes a
  Python 3 incompatibility.

Version 0.7.13
==============

- Drop explicit support for Python 2.6.

- Fix flake8 error.

Version 0.7.12
==============

- Avoid Python 3 exception ``configparser.DuplicateOptionError`` when reading
  ``~/.subversion/config`` having auto-props differing only in case, e.g.::

    [auto-props]
    *.c = svn:eol-style=native
    *.C = svn:eol-style=native

Version 0.7.11
==============

- Drop support for Python 3.3 (thanks to John Hagen).

Version 0.7.10
==============

- Add support for Python 3.7 (thanks to John Hagen).

Version 0.7.9
=============

- Add support for Python 3.6 (thanks to John Hagen).

- Document some overlooked switches::

    --no-pager           disable the automatic use of a pager
    --debug              enable debug printing (mainly for maintainer use)
    --svn path/to/svn    change path to ``svn`` utility (mainly for testing)

- Add ``--ie`` as an abbreviation for ``--ignore-externals``.

- Add ``svn pdiff`` command as a ``patch-compatible`` diff, equivalent to::

    svn diff --diff-cmd diff -x -U1000000 --patch-compatible

Version 0.7.8
=============

- Restore line suppression feature lost during the refactoring of the line
  generation logic.

Version 0.7.7
=============

- Make 'colorama' be a dependency.  This is necessary on Windows to support
  colorized output.

- Support additional Subversion switches:

  --force-interactive
  --ignore-properties
  --include-externals
  --no-diff-added
  --patch-compatible
  --properties-only
  --search
  --search-and
  --show-inherited-props

Version 0.7.6
=============

- Avoid DeprecationWarning for using SafeConfigParser in Python 3.2+.

- Avoid filtering stdout and stderr during "svn merge" operations.  This ensures
  that the output destined for the user's terminal comes out in the desired
  order; otherwise, it's possible for Subversion's menu choices to show up
  too early (e.g., when choosing the "diff-full" operation, the menu is
  redisplayed first followed by the actual diff output, leading to confusion).

- Change to unbuffered method of reading the "svn" client subprocess's output
  and error streams to provide realtime output display.

- Display total number of stderr characters instead of repeating all of stderr.

Version 0.7.5
=============

- Incorporate several code cleanups (thanks to John Hagen).

Version 0.7.4
=============

- Remove stray comma in description string in setup.py.

Version 0.7.3
=============

- Add Travis and PyPI badges to README (thanks to John Hagen).

- Allow full README to be shown on PyPI (thanks to John Hagen).

- Remove Python nightly version from Travis build (thanks to John Hagen).

Version 0.7.2
=============

- Document how to avoid URL adjustment (add ``/.`` to the end of the URL).

- Publish testsvn.py in the distribution.

- Provide initial support for Travis CI using py.test.

- Rename CHANGES.txt to CHANGES.rst, and LICENSE.txt to LICENSE.rst.

- Include Makefile and requirements.txt in MANIFEST.in.

Version 0.7.1
=============

- Add ``--strict`` to ``svn pge`` to avoid the spurious extra newline.

Version 0.7.0
=============

- Capture stderr from "svn" client, highlight these lines, and repeat them
  at the end of the operation so they are not overlooked.  To facilitate this,
  the method of restoring stdout (and now stderr as well) when invoking
  SVN_EDITOR was changed.  When invoking the editor, stdout and stderr are
  now connected to the platform-specific console device ('/dev/tty' on Unix
  machines, and 'CON:' on Windows).

- Add "pge" and "pgi" shortcuts to propget svn:externals and svn:ignores.

- Support svn v1.7+ svn:externals diff format.

  Newer svn clients now provide line-by-line diff output for changes to
  svn:externals, so svnwrap now detects this case to prevent erroneous
  formatting.

- Port to run on both Python 2.x and 3.x, with 2.6 as the minimum supported
  version of Python.

- Format for PEP8 compliance.

.. vim:set ft=rst:


test/test_svnwrap.py
#!/usr/bin/env python

import unittest
import svnwrap
import warnings


class TestSvnwrap(unittest.TestCase):
    def setUp(self):
        warnings.simplefilter("error")

    def test_svnwrap_config(self):
        svnwrap.svnwrap_config()

    def test_paths_equal(self):
        self.assertTrue(svnwrap.paths_equal("somepath", "somepath"))
        self.assertFalse(svnwrap.paths_equal("somepath1", "somepath2"))


if __name__ == "__main__":
    unittest.main()


tests/test_svnwrap.py
#!/usr/bin/env python
# coding=utf-8

try:
    from typing import (
        List,
    )
except ImportError:
    pass

import svnwrap


def test_svnwrap_config():
    # type: () -> None
    svnwrap.svnwrap_config()


def test_paths_equal():
    # type: () -> None
    assert svnwrap.paths_equal("somepath", "somepath")
    assert not svnwrap.paths_equal("somepath1", "somepath2")


def test_parse_switch():
    # type: () -> None

    def parse_zero(switch):
        # type: (str) -> None
        args = ["extra"]  # type: List[str]
        assert svnwrap.parse_switch(switch, args) == [switch]
        assert args == ["extra"]
        if not switch.startswith("--"):
            args = []
            assert svnwrap.parse_switch(switch + "f", args) == [switch]
            assert args == ["-f"]

    def parse_one(cmd):
        # type: (str) -> None
        switch, arg = cmd.split()
        args = [arg, "extra"]  # type: List[str]
        assert svnwrap.parse_switch(switch, args) == [switch, arg]
        assert args == ["extra"]

        args = ["extra"]
        if switch.startswith("--"):
            cmd = switch + "=" + arg
        else:
            cmd = switch + arg
        assert svnwrap.parse_switch(cmd, args) == [switch, arg]
        assert args == ["extra"]

    parse_zero("-N")
    parse_zero("-q")
    parse_zero("--version")
    parse_zero("--verbose")
    parse_one("--depth files")
    parse_one("-F file")


fakesvn.py
#!/usr/bin/env python
# coding=utf-8

import sys


def write(s):
    sys.stdout.write(s)
    sys.stdout.flush()


def write_ln(s):
    write(s + "\n")


def fake_status():
    write(
        """\
A      .
?      svnwrap.py
Performing status

X yadda

A      svnwrap.sh

?      Wiki-20
?      Account-Manager
C      svnwrap/fakesvn.py
G      svnwrap/merged.py
?      pyrepl
?      svnwrap/svnwrap.py

A      svnwrap
D      svnwrap/junkfile
A      svnwrap/svnwrap.sh
A      other/stuff.py
M      other/goodStuff.py
"""
    )


def fake_update():
    write(
        """\

Fetching external

External
?      svnwrap
A      svnwrap/svnwrap.sh
C      svnwrap/fakesvn.py
G      svnwrap/merged.py
A      other/stuff.py
M      other/goodStuff.py
Updated external

At revision 3.
"""
    )


def fake_switch():
    write(
        """\
A      svnwrap/svnwrap.sh
C      svnwrap/fakesvn.py
G      svnwrap/merged.py
D      other/stuff.py

At revision 2.
Updated to revision 2.
"""
    )


def fake_diff():
    write(
        """\

Property changes on: .
___________________________________________________________________
Name: svn:externals
   - # Section one
one file:///home/mrhenr1/tmp/svn/repo/trunk/one
two file:///home/mrhenr1/tmp/svn/repo/trunk/two

# Gap above, starting section two
three file:///home/mrhenr1/tmp/svn/repo/trunk/three

four file:///home/mrhenr1/tmp/svn/repo/trunk/four

   + # Section one
one file:///home/mrhenr1/tmp/svn/repo/trunk/one
five file:///home/mrhenr1/tmp/svn/repo/trunk/five

# Gap above, starting section two
three file:///home/mrhenr1/tmp/svn/repo/trunk/three

six file:///home/mrhenr1/tmp/svn/repo/trunk/six
# Extra line


Modified: svn:externals
## -1 +1,3 ##
+../comp0 comp0
 ../comp comp
+../comp3 comp3
Added: some_weird_property
+property_value goes here
Index: testfile.txt
===================================================================
--- testfile.txt        (revision 0)
+++ testfile.txt        (revision 0)
@@ -0,0 +1,4 @@
+This
+is
+a
+test file.

Property changes on: testfile.txt
___________________________________________________________________
Name: svn:eol-style
   + native

"""
    )


def fake_revert(args):
    write_ln("Reverting:")
    for f in args:
        write_ln("%r" % f)


def fake_log():
    write(
        """\
------------------------------------------------------------------------
r15 | committer | 2014-10-12 15:34:35 -0400 (Sun, 12 Oct 2014) | 4 lines

Even more bar.

And a multi-line log.

------------------------------------------------------------------------
r14 | committer | 2014-10-11 04:43:18 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r13 | committer | 2014-10-11 04:43:17 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r12 | committer | 2014-10-11 04:43:16 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r11 | committer | 2014-10-11 04:43:15 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r10 | committer | 2014-10-11 04:43:14 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r9 | committer | 2014-10-11 04:43:13 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r8 | committer | 2014-10-11 04:43:12 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r7 | committer | 2014-10-11 04:43:11 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r6 | committer | 2014-10-11 04:43:10 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r5 | committer | 2014-10-11 04:43:09 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r4 | committer | 2014-10-11 04:43:08 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r3 | committer | 2014-10-11 04:42:52 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r2 | committer | 2014-10-11 04:42:36 -0400 (Sat, 11 Oct 2014) | 1 line

add bar
------------------------------------------------------------------------
r1 | committer | 2014-10-11 04:42:10 -0400 (Sat, 11 Oct 2014) | 1 line

add foo
------------------------------------------------------------------------
"""
    )


def main():
    args = sys.argv[1:]

    if args:
        cmd = args.pop(0)
    else:
        cmd = ""

    if cmd in ["st", "status"]:
        fake_status()

    elif cmd in ["up", "update"]:
        fake_update()

    elif cmd in ["sw", "switch"]:
        fake_switch()

    elif cmd == "diff":
        fake_diff()

    elif cmd == "revert":
        fake_revert(args)

    elif cmd == "log":
        fake_log()

    elif cmd == "":
        write_ln("Type 'svn help' for usage.")

    else:
        write_ln("fakesvn unknown command: svn %s %s" % (cmd, " ".join(args)))


if __name__ == "__main__":
    main()


testsvn.py
#!/usr/bin/env python
# coding=utf-8

import sys


def write(s):
    sys.stdout.write(s)
    sys.stdout.flush()


def write_ln(s):
    write(s + "\n")


def fake_status():
    write(
        """\
A      .
?      svnwrap.py
Performing status

X yadda

A      svnwrap.sh

?      Wiki-20
?      Account-Manager
C      svnwrap/testsvn.py
G      svnwrap/merged.py
?      pyrepl
?      svnwrap/svnwrap.py

A      svnwrap
D      svnwrap/junkfile
A      svnwrap/svnwrap.sh
A      other/stuff.py
M      other/goodStuff.py
"""
    )


def fake_update():
    write(
        """\

Fetching external

External
?      svnwrap
A      svnwrap/svnwrap.sh
C      svnwrap/testsvn.py
G      svnwrap/merged.py
A      other/stuff.py
M      other/goodStuff.py
Updated external

At revision 3.
"""
    )


def fake_switch():
    write(
        """\
A      svnwrap/svnwrap.sh
C      svnwrap/testsvn.py
G      svnwrap/merged.py
D      other/stuff.py

At revision 2.
Updated to revision 2.
"""
    )


def fake_diff():
    write(
        """\

Property changes on: .
___________________________________________________________________
Name: svn:externals
   - # Section one
one file:///home/mrhenr1/tmp/svn/repo/trunk/one
two file:///home/mrhenr1/tmp/svn/repo/trunk/two

# Gap above, starting section two
three file:///home/mrhenr1/tmp/svn/repo/trunk/three

four file:///home/mrhenr1/tmp/svn/repo/trunk/four

   + # Section one
one file:///home/mrhenr1/tmp/svn/repo/trunk/one
five file:///home/mrhenr1/tmp/svn/repo/trunk/five

# Gap above, starting section two
three file:///home/mrhenr1/tmp/svn/repo/trunk/three

six file:///home/mrhenr1/tmp/svn/repo/trunk/six
# Extra line


Modified: svn:externals
## -1 +1,3 ##
+../comp0 comp0
 ../comp comp
+../comp3 comp3
Added: some_weird_property
+property_value goes here
Index: testfile.txt
===================================================================
--- testfile.txt        (revision 0)
+++ testfile.txt        (revision 0)
@@ -0,0 +1,4 @@
+This
+is
+a
+test file.

Property changes on: testfile.txt
___________________________________________________________________
Name: svn:eol-style
   + native

"""
    )


def fake_revert(args):
    write_ln("Reverting:")
    for f in args:
        write_ln("%r" % f)


def fake_log():
    write(
        """\
------------------------------------------------------------------------
r15 | committer | 2014-10-12 15:34:35 -0400 (Sun, 12 Oct 2014) | 4 lines

Even more bar.

And a multi-line log.

------------------------------------------------------------------------
r14 | committer | 2014-10-11 04:43:18 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r13 | committer | 2014-10-11 04:43:17 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r12 | committer | 2014-10-11 04:43:16 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r11 | committer | 2014-10-11 04:43:15 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r10 | committer | 2014-10-11 04:43:14 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r9 | committer | 2014-10-11 04:43:13 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r8 | committer | 2014-10-11 04:43:12 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r7 | committer | 2014-10-11 04:43:11 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r6 | committer | 2014-10-11 04:43:10 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r5 | committer | 2014-10-11 04:43:09 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r4 | committer | 2014-10-11 04:43:08 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r3 | committer | 2014-10-11 04:42:52 -0400 (Sat, 11 Oct 2014) | 1 line

more bar
------------------------------------------------------------------------
r2 | committer | 2014-10-11 04:42:36 -0400 (Sat, 11 Oct 2014) | 1 line

add bar
------------------------------------------------------------------------
r1 | committer | 2014-10-11 04:42:10 -0400 (Sat, 11 Oct 2014) | 1 line

add foo
------------------------------------------------------------------------
"""
    )


def main():
    args = sys.argv[1:]

    if args:
        cmd = args.pop(0)
    else:
        cmd = ""

    if cmd in ["st", "status"]:
        fake_status()

    elif cmd in ["up", "update"]:
        fake_update()

    elif cmd in ["sw", "switch"]:
        fake_switch()

    elif cmd == "diff":
        fake_diff()

    elif cmd == "revert":
        fake_revert(args)

    elif cmd == "log":
        fake_log()

    elif cmd == "":
        write_ln("Type 'svn help' for usage.")

    else:
        write_ln("testsvn unknown command: svn %s %s" % (cmd, " ".join(args)))


if __name__ == "__main__":
    main()
